### PR TITLE
Pirate mode!!!

### PR DESCRIPTION
**Pull request by nrathbone-turing**
Pushing changes to be merged after fixing locally some of the issues we encountered testing previously regarding `game#display_boards` and `player#place_ships_randomly`
- Was able to fix the issue that prevented player ship from displaying with all Xs when sunk from project evaluation debugging, too

Added `#pirate_mode` to give our game a bunch of personality and flair! This includes creating a bunch of dialogue options where entering "9" during ship selection activates `#pirate_mode`
- In pirate mode, the computer speaks with pirate dialogue, and additional flair is added throughout the game, including randomized pirate quotes for actions like starting turns, ending the game, and quitting, so it will be different-ish each time the game is played

**Note:** need to revisit tests based on simplecov report since now with all the additional dialogue options added, it's much lower (I should have known better and written all of the pirate banter and quotes tests first, so that's next on the list"

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -61,8 +59,10 @@ def valid_placement?(ship, coordinates)
         else
             return false 
         end
-        return false if coordinates.any? { |coordinate| @cells[coordinate].empty? == false }
 
+        return false if coordinates.any? { |coordinate| !@cells[coordinate]&.empty? }
```
Comment (nrathbone-turing) [inline lib/board.rb:63]: added this safeguard to prevent infinite loops during testing and refactored the statement to be falsey

File: lib/cell.rb [hunk 2]
```diff
@@ -21,8 +21,11 @@ def fired_upon?
   end
 
   def fire_upon
-    # added this to prevent double damage being possible
-    return if @fired_upon
+    
+    if @fired_upon
+      puts "You've already fired there!"
```
Comment (nrathbone-turing) [inline lib/cell.rb:26]: added a printed message to confirm that the coordinate chosen had already been fired upon

File: lib/game.rb [hunk 3]
```diff
@@ -31,24 +47,32 @@ def start_game
...
     exit
   end
 
   def setup_game
-    # add sassy personality flair here for computer maybe?
     ship_options = {
       "1" => { name: "Cruiser", length: 3 },
-      "2" => { name: "Submarine", length: 2 }
+      "2" => { name: "Submarine", length: 2 },
+      "9" => { name: "Black Pearl", length: 4, hidden: true }
```
Comment (nrathbone-turing) [inline lib/game.rb:57]: here's the easter egg I added -- by adding it as a hidden option, this way the rest of the logic that only allows the player to select a total number of ships they wish to play with that cannot be greater than the number of ships displayed, but they can still enter "9" to enter pirate mode!

File: lib/game.rb [hunk 4]
```diff
@@ -31,24 +47,32 @@ def start_game
...
-    puts "Choose the ships you'll play with:\n\n"
+    if pirate_mode
+      puts "Choose yer vessels, ye salty dog:\n\n"
+    else
+      puts "Choose the ships you'll play with:\n\n"
+    end
 
     ship_options.each do |key, ship_info|
-      puts "#{key}. #{ship_info[:name]} (#{ship_info[:length]} spaces)"
+      ship_name = @pirate_mode ? pirate_ship_name(ship_info[:name]) : ship_info[:name]
```
Comment (nrathbone-turing) [inline lib/game.rb:67]: Need to refactor this one to be clearer about what it's doing and how it's checking for pirate mode, since this is the logic that renames the ship choices to piratey names if the game is retarted with pirate mode enabled

File: lib/game.rb [hunk 5]
```diff
@@ -60,21 +84,38 @@ def setup_game
...
+        end
       end
     end
 
     chosen_ships = []
 
     ship_count.times do |i|
       print "Enter the number for Ship #{i + 1}: "
-      user_input = gets.chomp
+      user_input = gets.chomp.strip
```
Comment (nrathbone-turing) [inline lib/game.rb:99]: added this because it worked well with the coordinates too to prevent invalid inputs

File: lib/player.rb [hunk 6]
```diff
@@ -79,25 +119,49 @@ def generate_valid_random_coordinates(length)
...
-
-        until placed
-            coordinates = generate_valid_random_coordinates(ship.length)
-            if board.valid_placement?(ship, coordinates)
-                board.place_ship(ship, coordinates)
-                placed = true
-                @ships << ship
-            end            
+        attempts = 0
+        max_attempts = 100
```
Comment (nrathbone-turing) [inline lib/player.rb:123]: added a max of 100 attempts here to act as a way to stub or limit the infinite loops that kept happening with `player#place_ship_randomly` because the random logic was working too well and would internally run many times before coming up with a valid input option, tried smaller numbers for max_attempts too including samples of 5 or 20, but 100 seemed to be the most consistent

File: lib/player.rb [hunk 7]
```diff
@@ -79,25 +119,49 @@ def generate_valid_random_coordinates(length)
...
+          if board.valid_placement?(ship, coordinates)
+            board.place_ship(ship, coordinates)
+            @ships << ship
+            return
+          end
+          attempts += 1
         end
-    end
+      
+        puts "DEBUG: Could not place ship #{ship.name} after #{max_attempts} attempts."
```
Comment (nrathbone-turing) [inline lib/player.rb:135]: added a specific debug statement to show us where and which attempt it was that the loop was getting to before crashing or causing us to force quit from the test which is partially how I determined what value to set for `max_attempts` above

File: play_battleship.rb [hunk 8]
```diff
@@ -4,5 +4,7 @@
 require './lib/game'
 require './lib/player'
 
-game = Game.new
-game.start_game
\ No newline at end of file
+if __FILE__ == $PROGRAM_NAME
```
Comment (nrathbone-turing) [inline play_battleship.rb:7]: spent a long time searching online for this one but ultimately I needed a way to stop getting prompted to play the game again after each test, so I was able to find a way to indicate that the runner file itself should only execute the code to start a new game when it's being run directly from the command line or terminal and not in any RSpec or other tests

File: spec/board_spec.rb [hunk 9]
```diff
@@ -52,6 +52,21 @@
             expect(board.valid_placement?(cruiser, ["A3", "A2", "A1"])).to eq(false)
             expect(board.valid_placement?(submarine, ["C1", "B1"])).to eq(false)
         end
+
+        it 'is false if coordinates are diagonal' do
+            board = Board.new
+            cruiser = Ship.new("Cruiser", 3)
```
Comment (nrathbone-turing) [inline spec/board_spec.rb:58]: added additional tests for edge cases like diagonal placements and duplicate coordinates to return invalid message

File: spec/board_spec.rb [hunk 10]
```diff
@@ -85,13 +101,24 @@
...
+    describe 'is rendering board correctly based on various game actions and board states' do
        
         before(:each) do
             @board = Board.new
             @cruiser = Ship.new("Cruiser", 3)
         end
-        it 'in a  4x4 grid formation with cells containing ships hidden by default' do
+
+        it 'returns a string from render' do
+            expect(@board.render).to be_a(String)
```
Comment (nrathbone-turing) [inline spec/board_spec.rb:112]: based on simplecov score for this file, added some more tests to confirm the render was a string as expected, and that the rows and columns were correct for what was expected in terms of letters and numbers, respectively

File: spec/board_spec.rb [hunk 11]
```diff
@@ -122,23 +149,69 @@
 
             expect(@board.render(true)).to eq(expected_render)
         end
-    end
 
+        it 'renders a board with a missed shot' do
+            @board.cells["B4"].fire_upon
+            
+            expected_render = 
+                "  1 2 3 4 \n" +
```
Comment (nrathbone-turing) [inline spec/board_spec.rb:157]: added additional tests for more advanced board states in the renders

File: spec/board_spec.rb [hunk 12]
```diff
@@ -122,23 +149,69 @@
...
+              "B X . . M \n" +
+              "C X . . . \n" +
+              "D X . . . \n"
+          
+            expect(@board.render).to eq(expected_render)
+        end
+
+        it 'can correctly render a ship that is hit but not sunk with show_ship == true' do
+            @board.place_ship(@cruiser, ["A1", "A2", "A3"])
+            @board.cells["A1"].fire_upon
```
Comment (nrathbone-turing) [inline spec/board_spec.rb:209]: i wasn't exactly sure how to write a test for this one and/or if we needed multiple tests to re-create the board state we encountered during the eval live demo, where the computer sunk one of the player's ships but it showed S still instead of X, but only for one of the ships that was placed, but this was my attempt to check for an include statement instead of a specific board state / render

File: spec/cell_spec.rb [hunk 13]
```diff
@@ -107,12 +107,10 @@
 
     it 'returns "X" if the cell has been fired upon and its ship has sunk' do
       @cell_2.place_ship(@cruiser)
-      @cell_2.fire_upon
-      
       3.times {@cruiser.hit}
+      @cell_2.fire_upon
```
Comment (nrathbone-turing) [inline spec/cell_spec.rb:111]: had to swap the order here so that the fire_upon method was called after the ship had 0 health to get this test to pass

File: spec/game_spec.rb [hunk 14]
```diff
@@ -29,32 +29,29 @@
           expect(@game.welcome_message).to eq("Welcome to BATTLESHIP\nEnter 'p' to play. Enter 'q' to quit.")
         end
 
-        # refactor notes - commenting these out for now because we don't have to test the user inputs,
-        # but f there's time left, this could be something to try mocks and stubs with to create a fake input/method
         it 'allows the user to choose to play by typing "p"' do
-          # expect(@game.user_input = 'p').to eq(true)
+          expect(@game.user_input = 'p').to output(true).stdout
```
Comment (nrathbone-turing) [inline spec/game_spec.rb:33]: changed the format of these expect statements to try and test the terminal outputs or user inputs even though we did not need to; I don't know that I believe these tests are passing and/or are correct, but wanted to give it a try/discuss if time allowed

File: spec/game_spec.rb [hunk 15]
```diff
@@ -91,23 +88,16 @@
           @player2 = Player.new("player")
           @player1.computer_player
 
-          # expected_render = 
-          # "=============COMPUTER BOARD============="
-          # @player1.board.render
-          # "==============PLAYER BOARD=============="
-          # @player2.board.render(true)
+          expected_render = 
+            "=============COMPUTER BOARD=============\n" +
```
Comment (nrathbone-turing) [inline spec/game_spec.rb:92]: Refactored the way the expected_render was returning for the test both in terms of the terminal output but because of the scary deprecation error message (https://turingschool.slack.com/archives/C08H85CL32R/p1744943380411589)

Comment (nrathbone-turing) [reply spec/game_spec.rb:92]: Brady was saying to do it more like this though:

`expect(board.render(true)).to eq("A B C D /n...")`

which he was saying the runner file would then execute `puts board.render(true)` to output to terminal, which isn't tested but still verified in a sense
